### PR TITLE
refactor: migrate from Gson to KotlinX.Serialization

### LINKED ISSUE #305: Migrate from Gson to KotlinX.Serialization
The project is currently using Gson for JSON serialization, which is not fully compatible with Kotlin Multiplatform. We've successfully migrated CaptureResult to KotlinX.Serialization in iOS module, but we still need to migrate CaptureResults and other related classes.
https://github.com/takahirom/roborazzi/blob/main/roborazzi-compose-ios/src/iosMain/kotlin/io/github/takahirom/roborazzi/CaptureResult.kt

### PR DESCRIPTION
**Pull request by sanao1006**
## Overview
Migrate json library to kotlinx.serialization since Gson does not support KMM.

## issue
Fixes #305

### CODE REVIEW

**Comment by takahirom** [comment]
Looks great. It's a bit challenging, but there seems to be an error.

```
Caused by: kotlinx.serialization.json.internal.JsonDecodingException: Expected start of the object '{', but had 'EOF' instead at path: $
JSON input: {"golden_file_path":"/Users/.....
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail(AbstractJsonLexer.kt:598)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail$default(AbstractJsonLexer.kt:596)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail$kotlinx_serialization_json(AbstractJsonLexer.kt:233)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail$kotlinx_serialization_json$default(AbstractJsonLexer.kt:228)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.unexpectedToken(AbstractJsonLexer.kt:225)
	at kotlinx.serialization.json.internal.StringJsonLexer.consumeNextToken(StringJsonLexer.kt:77)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.beginStructure(StreamingJsonDecoder.kt:102)
	at com.github.takahirom.roborazzi.CaptureResult$Unchanged$$serializer.deserialize(CaptureResult.kt:87)
	at com.github.takahirom.roborazzi.CaptureResult$Unchanged$$serializer.deserialize(CaptureResult.kt:87)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at com.github.takahirom.roborazzi.CaptureResult$CaptureResultSerializer.deserialize(CaptureResult.kt:129)
	at com.github.takahirom.roborazzi.CaptureResult$CaptureResultSerializer.deserialize(CaptureResult.kt:109)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at kotlinx.serialization.json.Json.decodeFromString(Json.kt:107)
	at com.github.takahirom.roborazzi.CaptureResult$Companion.fromJsonFile(CaptureResult.kt:137)
	at io.github.takahirom.roborazzi.RoborazziPlugin$apply$configureRoborazziTasks$finalizeTestRoborazziTask$1$execute$2.invoke(RoborazziPlugin.kt:276)
	at io.github.takahirom.roborazzi.RoborazziPlugin$apply$configureRoborazziTasks$finalizeTestRoborazziTask$1$execute$2.invoke(RoborazziPlugin.kt:246)
	at io.github.takahirom.roborazzi.RoborazziPlugin$apply$configureRoborazziTasks$finalizeTestRoborazziTask$1.execute$lambda$1(RoborazziPlugin.kt:246)
```

**Comment by sanao1006** [comment]
Sorry, do you have any idea as to the cause of the CI error?

I'm a little stuck and need some help...

**Comment by takahirom** [comment]
It is close. The integration test says this 👀 

```
kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 446: Expected start of the object '{', but had ']' instead at path: $.results[0]
JSON input: .....razziTest"},"type":"recorded"}]}
	at app//kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
	at app//kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.fail(AbstractJsonLexer.kt:598)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.fail$default(AbstractJsonLexer.kt:596)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.fail$kotlinx_serialization_json(AbstractJsonLexer.kt:233)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.fail$kotlinx_serialization_json$default(AbstractJsonLexer.kt:228)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.unexpectedToken(AbstractJsonLexer.kt:225)
	at app//kotlinx.serialization.json.internal.StringJsonLexer.consumeNextToken(StringJsonLexer.kt:74)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.beginStructure(StreamingJsonDecoder.kt:102)
	at app//com.github.takahirom.roborazzi.CaptureResult$Recorded$$serializer.deserialize(CaptureResult.kt:40)
	at app//com.github.takahirom.roborazzi.CaptureResult$Recorded$$serializer.deserialize(CaptureResult.kt:40)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at app//com.github.takahirom.roborazzi.CaptureResult$CaptureResultSerializer.deserialize(CaptureResult.kt:129)
	at app//com.github.takahirom.roborazzi.CaptureResult$CaptureResultSerializer.deserialize(CaptureResult.kt:111)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at app//kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
	at app//kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableElement(StreamingJsonDecoder.kt:168)
	at app//kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:538)
	at app//kotlinx.serialization.internal.CollectionLikeSerializer.readElement(CollectionSerializers.kt:80)
	at app//kotlinx.serialization.internal.AbstractCollectionSerializer.readElement$default(CollectionSerializers.kt:51)
	at app//kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:36)
	at app//kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:43)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at app//kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
	at app//kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableElement(StreamingJsonDecoder.kt:168)
	at app//com.github.takahirom.roborazzi.CaptureResults$$serializer.deserialize(CaptureResults.kt:21)
	at app//com.github.takahirom.roborazzi.CaptureResults$$serializer.deserialize(CaptureResults.kt:21)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at app//kotlinx.serialization.json.Json.decodeFromString(Json.kt:107)
	at app//com.github.takahirom.roborazzi.CaptureResults$Companion.fromJsonFile(CaptureResults.kt:222)
	at app//io.github.takahirom.roborazzi.RoborazziGradleProject.checkResultCount(RoborazziGradleProject.kt:320)
	at app//io.github.takahirom.roborazzi.RoborazziGradleProject.checkResultCount$default(RoborazziGradleProject.kt:312)
	at app//io.github.takahirom.roborazzi.RoborazziGradleProjectTest.shouldNotRetainPreviousTestResults(RoborazziGradleProjectTest.kt:451)
	at java.base@11.0.23/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.23/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.23/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.23/java.lang.reflect.Method.invoke(Method.java:566)
	at app//org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at app//org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at app//org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at app//org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at app//org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at app//org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at app//org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at app//org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at app//org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at app//org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at app//org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at app//org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at app//org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at app//org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at app//org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at app//org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at app//org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:112)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:40)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:60)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:52)
	at java.base@11.0.23/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.23/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.23/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.23/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
verify_changeDetect
kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 781: Expected start of the object '{', but had ']' instead at path: $.results[0]
JSON input: .....ext_data":{},"type":"changed"}]}
	at app//kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
	at app//kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.fail(AbstractJsonLexer.kt:598)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.fail$default(AbstractJsonLexer.kt:596)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.fail$kotlinx_serialization_json(AbstractJsonLexer.kt:233)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.fail$kotlinx_serialization_json$default(AbstractJsonLexer.kt:228)
	at app//kotlinx.serialization.json.internal.AbstractJsonLexer.unexpectedToken(AbstractJsonLexer.kt:225)
	at app//kotlinx.serialization.json.internal.StringJsonLexer.consumeNextToken(StringJsonLexer.kt:74)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.beginStructure(StreamingJsonDecoder.kt:102)
	at app//com.github.takahirom.roborazzi.CaptureResult$Changed$$serializer.deserialize(CaptureResult.kt:72)
	at app//com.github.takahirom.roborazzi.CaptureResult$Changed$$serializer.deserialize(CaptureResult.kt:72)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at app//com.github.takahirom.roborazzi.CaptureResult$CaptureResultSerializer.deserialize(CaptureResult.kt:130)
	at app//com.github.takahirom.roborazzi.CaptureResult$CaptureResultSerializer.deserialize(CaptureResult.kt:111)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at app//kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
	at app//kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableElement(StreamingJsonDecoder.kt:168)
	at app//kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:538)
	at app//kotlinx.serialization.internal.CollectionLikeSerializer.readElement(CollectionSerializers.kt:80)
	at app//kotlinx.serialization.internal.AbstractCollectionSerializer.readElement$default(CollectionSerializers.kt:51)
	at app//kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:36)
	at app//kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:43)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at app//kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
	at app//kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableElement(StreamingJsonDecoder.kt:168)
	at app//com.github.takahirom.roborazzi.CaptureResults$$serializer.deserialize(CaptureResults.kt:21)
	at app//com.github.takahirom.roborazzi.CaptureResults$$serializer.deserialize(CaptureResults.kt:21)
	at app//kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at app//kotlinx.serialization.json.Json.decodeFromString(Json.kt:107)
	at app//com.github.takahirom.roborazzi.CaptureResults$Companion.fromJsonFile(CaptureResults.kt:222)
	at app//io.github.takahirom.roborazzi.RoborazziGradleProject.checkResultCount(RoborazziGradleProject.kt:320)
	at app//io.github.takahirom.roborazzi.RoborazziGradleProject.checkResultCount$default(RoborazziGradleProject.kt:312)
	at app//io.github.takahirom.roborazzi.RoborazziGradleProjectTest.verify_changeDetect(RoborazziGradleProjectTest.kt:195)
```

File: include-build/roborazzi-core/src/jvmTest/kotlin/io/github/takahirom/roborazzi/CaptureResultTest.kt [hunk 1]
```diff
@@ -175,7 +180,7 @@ class CaptureResultTest {
     assertEquals(File("compare_file"), actualAddedResult.compareFile)
     assertEquals(File("actual_file"), actualAddedResult.actualFile)
     assertEquals(123456789, actualAddedResult.timestampNs)
-    assertEquals(2, (actualAddedResult.contextData["key"] as Double).toInt())
+    assertEquals(2, (actualAddedResult.contextData["key"].toString().toDouble()).toInt())
```
Comment (takahirom) [inline include-build/roborazzi-core/src/jvmTest/kotlin/io/github/takahirom/roborazzi/CaptureResultTest.kt:183]: 👀 

      "context_data": {
        "test_key": "test_value",
        "test_key2": "10"
      },

File: include-build/roborazzi-core/src/commonJvmMain/kotlin/com/github/takahirom/roborazzi/CaptureResults.kt [hunk 2]
```diff
@@ -196,3 +200,21 @@ data class CaptureResults(
...
   }
 }
+
+object AnySerializer : KSerializer<Any> {
+  @OptIn(ExperimentalSerializationApi::class)
+  override val descriptor: SerialDescriptor
+    get() = ContextualSerializer(Any::class, null, emptyArray()).descriptor
+
+  override fun serialize(encoder: Encoder, value: Any) {
+    encoder.encodeString(value.toString())
```
Comment (takahirom) [inline include-build/roborazzi-core/src/commonJvmMain/kotlin/com/github/takahirom/roborazzi/CaptureResults.kt:210]: I think you know, but we might have to make some changes here. 👀

File: sample-android/src/test/java/com/github/takahirom/roborazzi/sample/boxed/ContextDataTest.kt [hunk 3]
```diff
@@ -150,13 +150,12 @@ class ContextDataTest {
...
     val any = contextData[key]
     if (any is Double) {
-      // gson always parse number as double
       assert(any.toInt() == value.toInt()) {
         "Expected $value but got $any"
       }
       return
     }
-    assert(any == value) {
+    assert(any.toString() == value) {
```
Comment (sanao1006) [inline sample-android/src/test/java/com/github/takahirom/roborazzi/sample/boxed/ContextDataTest.kt:158]: This part seems a bit unnatural.
But without `toString()`, the assertion fails.
What do you think?

**Comment by takahirom** [comment]
I think we were able to improve the current JSON thanks to kotlinx.serialization!
I'll merge this PR after passing integration tests. Thank you for your work!

**Pull request merged**